skt/retry.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2018 Red Hat, Inc. All rights reserved. This copyrighted
+# material is made available to anyone wishing to use, modify, copy, or
+# redistribute it subject to the terms and conditions of the GNU General Public
+# License v.2 or later.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+"""Retry decorator to retry method/function several times."""
+
+import time
+import logging
+
+
+def retry(exception, retries=3, initial_delay=3):
+    """Decorator to retry method/function that raised exception after
+       initial_delay seconds. Do retries based on how many were set."""
+    def wrapper(function):
+        def wrapped(*args, **kwargs):
+            wrapped.failed_count = 0
+            wrapped.retries = retries
+
+            delay = 0
+            for i in range(0, retries):
+                delay += initial_delay
+
+                try:
+                    return function(*args, **kwargs)
+                except exception as exc:
+                    wrapped.failed_count += 1
+
+                    err_msg = str(exc) + ", delaying for %ds" % delay
+                    logging.warning(err_msg)
+                    for arg in args:
+                        logging.warning(str(arg))
+
+                    # don't sleep on last attempt, there's no point
+                    if i != retries - 1:
+                        time.sleep(delay)
+                    else:
+                        # reraise exc. instead
+                        raise exc
```

Comment:
I don't understand this: why "instead"? Instead of what?

Also, maybe related to my confusion: isn't it better to just use `raise` without parameters, to keep stack information and such?

Reply:
1) Look one or two lines above. :D Well, ideally it would be:
`
                    if i != retries - 1:
                        s = str(e) + ", delaying for %ds" % delay
                        logging.warning(s)
                        time.sleep(delay)
                    else:
                        # don't sleep on last attempt, there's no point;
                        # instead reraise exception
                        raise
`
Meaning "retry" is the most common case (print warning + wait), but on last attempt we reraise the exception instead(of waiting). Makes sense?

2) Yes, no doubt.